walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,79 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> spots = new ArrayList<TimeRange>();
+    ArrayList<String> attendees = new ArrayList<String>(request.getAttendees());
+    ArrayList<Event> eventsList = new ArrayList<Event>(events);
+    long duration = request.getDuration();
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return spots;
+    }
+    
+    if (eventsList.isEmpty() || attendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    /** Sorts the events list by start time. */
+    boolean sorted = true;
+    do {
+      for (int i = 0; i < eventsList.size() - 1; i++) {
+        if (TimeRange.ORDER_BY_START
+            .compare(eventsList.get(i).getWhen(), eventsList.get(i+1).getWhen()) > 0) {
+          Event tmp = eventsList.get(i);
+          eventsList.set(i, eventsList.get(i+1));
+          eventsList.set(i+1, tmp);
+          sorted = false;
+        }
+      }
+    } while (!sorted);
+
+    /** Deletes irrelevant events and events contained by other events. */
+    for (int i = 0; i < eventsList.size(); i++) {
+      boolean relevant = false;
+      int j = 0;
+      while (j < attendees.size() && (!relevant)) {
+        if (eventsList.get(i).getAttendees().contains(attendees.get(j))) {
+          relevant = true;
+        }
+        j++;
+      }
+      if (!relevant) {
+        eventsList.remove(i);
+        i--;
+      } else if ((i < eventsList.size() - 1) && 
+                 (eventsList.get(i).getWhen().contains(eventsList.get(i+1).getWhen()))) {
+        eventsList.remove(i+1);
+        i--;
+      }
+    }
+
+    if (eventsList.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    int start = TimeRange.START_OF_DAY;
+    int end = eventsList.get(0).getWhen().start();
+    int spotDuration = end - start;
+    for (int i = 0; i < eventsList.size(); i++) {
+      if (spotDuration >= duration) {
+        spots.add(TimeRange.fromStartEnd(start, end, false));
+      }
+      start = eventsList.get(i).getWhen().end();
+      end = (i == eventsList.size() - 1) ? TimeRange.END_OF_DAY : eventsList.get(i+1).getWhen().start();
+      spotDuration = end - start;
+    }
+    // From the end of the last event to the end of day
+    if (spotDuration >= duration) {
+      spots.add(TimeRange.fromStartEnd(start, end, true));
+    }
+
+    return spots;
```

Comment:
please consider refactoring like:
```
int start = TimeRange.START_OF_DAY;
for (event : eventsList) {
  end = event.getWhen().end();
  if (end - start >= duration) {
    spots.add(TimeRange.fromStartEnd(start, end, false));
  }
  start = max(start, end);
}
if (TimeRange.END_OF_DAY - start >= duration) {
  spots.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, false));
}
```
I think the code above is clearer.

Reply:
IIUC, the code above compares the request duration to the time between the ends of two events, which includes both the available gap between events and the length of the next event itself. This might potentially create incorrect results. I'll try to refactor the code to make it clearer.